File: examples/extract.py
#!/usr/bin/env python3
"""extract.py template: decode a PNG file to a payload file.
Usage: extract.py <image.png> <out.bin> [--password P]
Decode only. What you DO with the bytes afterwards is your responsibility.
"""
import sys

from PIL import Image

sys.path.insert(0, __import__('os').path.join(
    __import__('os').path.dirname(__file__), '..', 'python'))
from stegolib import decode_image


def main():
    password = ''
    args = [a for a in sys.argv[1:] if not a.startswith('--password')]
    for i, a in enumerate(sys.argv[1:]):
        if a == '--password' and i + 1 < len(sys.argv[1:]):
            password = sys.argv[1:][i + 1]
    if len(args) != 2:
        print('usage: extract.py <image.png> <out.bin> [--password P]')
        return 2
    img = Image.open(args[0]).convert('RGB')
    w, h = img.size
    flat = []
    for r, g, b in img.getdata():
        flat += [r, g, b]
    payload = decode_image(flat, w, h, password=password)
    if payload is None:
        print('decode failed')
        return 1
    open(args[1], 'wb').write(payload)
    print(f'decoded {len(payload)} bytes')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: examples/extract.rs
// extract.rs — template: decode an image file to a payload file.
// Usage: extract <image.png> <out.bin> [password]
// Decode only. What you DO with the bytes afterwards is your responsibility.
//
// PNG loading: `image` crate (add `image = "0.25"` to your own Cargo.toml;
// kept out of the library so the template stays dependency-explicit).
use std::env;
use stego::{decode, StegoError};

fn load_rgb(path: &str) -> Result<(Vec<u8>, u32, u32), String> {
    let img = image::open(path).map_err(|e| e.to_string())?.to_rgb8();
    let (w, h) = img.dimensions();
    Ok((img.into_raw(), w, h))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("usage: extract <image.png> <out.bin> [password]");
        std::process::exit(2);
    }
    let pw = args.get(3).map(|s| s.as_str());
    let (rgb, w, h) = load_rgb(&args[1]).unwrap_or_else(|e| {
        eprintln!("load failed: {e}");
        std::process::exit(2);
    });
    match decode(&rgb, w, h, pw) {
        Ok(payload) => {
            std::fs::write(&args[2], &payload).unwrap_or_else(|e| {
                eprintln!("write failed: {e}");
                std::process::exit(1);
            });
            println!("decoded {} bytes", payload.len());
        }
        Err(StegoError::Auth) => {
            eprintln!("decode failed: authentication/password");
            std::process::exit(1);
        }
        Err(e) => {
            eprintln!("decode failed: {e:?}");
            std::process::exit(1);
        }
    }
}
